ansible-galaxy - add configuration options for more flexible collection signature verification

s-hertel opened the pull request:
##### SUMMARY
Follow-up to #76681.

* Add a toggle to control the number of signatures required to verify the authenticity of a collection. The default is now "1" (rather than "all"), which is a change in behavior from #76681.

  This is a positive integer or `all` and may have a leading `+` to denote whether it should be an error if no signatures are found on the Galaxy server or were provided by the user. For example `ansible-galaxy collection install ns.coll --required-valid-signature-count +1` will error if no signatures are found (assuming the keyring is provided).

* Add a toggle to limit the gpg --verify status codes that are considered a failure

Still need to:
- [x] Bikeshed option names (they're pretty long)
       No other suggestions.
- [x] Bikeshed behavior if all errors are ignored but no signatures were truly successful
      Addressed by adding 'strict' signature verification. Use a positive integer/all preceded by `+` to require signature verification.
- [x] Add More tests
- [x] Add Documentation/update existing documentation
- [x] Add Changelog

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
Add a toggle to control the number of signatures required to verify the authenticity of a collection

Default to 1 instead of all (-1)

Add a toggle to limit the gpg status codes that are considered a failure

Add some initial integration tests

s-hertel committed:
Limit verify_file_signatures to if the keyring is configured and add FIXME to figure out

s-hertel committed:
Consolidate logic. Don't special-case no signatures.

Add changelog

Update documentation

Add more tests

s-hertel commented:
Here's a table to make it easier to see the current behavior for all combinations of factors: https://gist.github.com/s-hertel/aa08ecf4d44bb9578d815689f9d46118

I also added a unit test for those combinations.

nitzmahone commented:
I [commented on the table gist](https://gist.github.com/s-hertel/aa08ecf4d44bb9578d815689f9d46118?permalink_comment_id=4078390#gistcomment-4078390) - not sure if we want to discuss here or there...

s-hertel commented:
@nitzmahone Thanks, I left a response to your reply there as well (probably easier to follow in the context of that gist).

s-hertel committed:
Fixes #77146

Fix using user-provided signatures when running 'ansible-galaxy collection verify ns.coll --offline'

Add a test for a user-provided signature when running ansible-galaxy collection verify with --offline

Fix displaying overall gpg failure without extra verbosity

Add a test for displaying gpg failure without verbosity

Improve documentation to be more clear that signature verification only currently applies to collections directly sourced from Galaxy servers

s-hertel committed:
Add option to make signature verification strict and fail if there are no valid signatures

s-hertel committed:
update unit tests

s-hertel commented:
I updated the gist with a new table to show what 654d33cf4513126e296167d1b7e1f6ec2ba5cc14 (the addition of 'strict' signature verification) has changed, and left a comment on how those changes impact the main points of concern: https://gist.github.com/s-hertel/aa08ecf4d44bb9578d815689f9d46118?permalink_comment_id=4088633#gistcomment-4088633

samccann reviewed (approved):
Docs portion LGTM

nitzmahone commented on lib/ansible/config/base.yml:
```diff
@@ -1491,16 +1491,17 @@ GALAXY_IGNORE_INVALID_SIGNATURE_STATUS_CODES:
-  type: int
-  default: 1
+  type: str
+  default: 1  # should this be strict (+1) by default?
```
Tough call, but I'd say probably not, at least to start- even though we won't be supplying a keyring by default, we probably don't want the upstream version to default to everything must be signed if someone does supply a keyring. The product-packaged version might want to patch that at some point, though.

s-hertel replied:
That makes sense. I removed the comment, to be revisited later perhaps.

s-hertel committed:
Update type hint for required successful signature count to be a str. The value should be "all" or a string that is castable to a positive integer. Either may use the prefix "+" to fail if no signatures have been found to verify the collection.

Match devel type hinting

nitzmahone commented on lib/ansible/cli/galaxy.py:
```diff
@@ -146,6 +146,23 @@ def _get_collection_widths(collections):
+def validate_signature_count(value):
+    value = str(value)
+
+    if value.startswith('+'):
```
This could be more robust with something like `^(?P<strict>\+)?(?:(?P<count>\d+)|(?P<all>all))$` - then you can just ask for the various named bits on a match with full assurance that you've got something valid, and just make the error message on a non-match generic like `(bla) is not a valid signature count value`. As-is, there are several cases that would just blow up with a generic `ValueError` from `int`.

s-hertel replied:
Yeah, that's much better. Also fixes floats, which shouldn't be valid!

nitzmahone commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -329,41 +329,46 @@ def verify_local_collection(
     successful = 0
     error_messages = []
 
+    strict = required_successful_count.startswith('+')
```
If you did the regex thing above, you could probably also share this logic with the CLI validator

s-hertel replied:
Fixed. Thanks very much for the careful review, the state table gave me quite a headache :smile:

nitzmahone reviewed (changes requested):
Just a couple suggestions to tighten up the validation of the composite value, but I've gone cross-eyed comparing it to the last version of the state table (as I'm sure you have too ;) ) and it looks like it does what it's supposed to in all those cases. Thanks!

s-hertel committed:
Use a regex to validate --required-valid-signature-count

Add unit test for valid/invalid signature count options

Do + report signature verification as long as keyring is provided

s-hertel committed:
add integration tests for signature verification without signatures

nitzmahone reviewed (approved):
Yep, looks good!

s-hertel commented:
@bcoca @samccann @nitzmahone Thank you for the reviews and brainstorming!
